File: problemList/slidingWindow/solution/LC1234.java
package problemList.slidingWindow.solution;

import java.util.HashMap;
import java.util.Map;

/**
替换的子串越长, 越有可能使得替换后的字符串变成一个平衡字符串, 具有单调性, 可以使用滑窗
并且这是一个 越长越合法 的滑窗
 */
public class LC1234 {
    private static Map<Character, Integer> map = new HashMap<>();
    static{
        map.put('Q', 0);
        map.put('W', 1);
        map.put('E', 2);
        map.put('R', 3);
    }
    public int balancedString(String s) {
        int n = s.length(), left = 0, ret = Integer.MAX_VALUE;
        int[] cnt = new int[4];
        for(char c : s.toCharArray()) cnt[map.get(c)]++;
        if(check(cnt, 0)) return 0;
        for(int i = 0;i < n;i++){
            cnt[map.get(s.charAt(i))]--;
            while(left <= i && check(cnt, i - left + 1)){
                ret = Math.min(ret, i - left + 1);
                cnt[map.get(s.charAt(left))]++;
                left++;
            }
        }
        return ret;
    }
    private boolean check(int[] cnt, int i) {
        int max = 0, tot = 0;
        for(int x : cnt) {
            max = Math.max(max, x);
            tot += x;
        }
        if(3 * max <= tot + i - max) return true;
        else return false;
    }
}

File: SolutionsWithCorCpp/problem_list_0x3f/sliding_window/solution/LC1234.cpp
#include <algorithm>
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

// 只要对于选定的区间之外的元素来说, 没有一个元素的频率超过 n / 4 ,
// 那么总存在一种替换方式

int getIdx(char c) {
	if (c == 'Q') return 0;
	else if (c == 'W') return 1;
	else if (c == 'E') return 2;
	else if (c == 'R') return 3;
	else return -1;
}

bool check(int cnt[], int totCnt[], int n) {
	for (int i = 0; i < 4; i++) {
		if (totCnt[i] - cnt[i] > n / 4) return false;
	}
	return true;
}

int balancedString(string s) {
	int n = s.size();
	int totCnt[4] = {0};
	for (char c : s) {
		totCnt[getIdx(c)]++;
	}
	int cnt[4] = {0};
	int r = 0, ret = n;
	for (int i = 0; r < n; r++) {
		cnt[getIdx(s[r])]++;
		while (check(cnt, totCnt, n) && i < n) {
			ret = min(ret, r - i + 1);
			cnt[getIdx(s[i])]--;
			i++;
		}
	}
	return ret;
}